File: android-core/src/androidTest/java/com/mparticle/networking/PinningTestHelper.java
package com.mparticle.networking;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.mparticle.identity.MParticleIdentityClient;
import com.mparticle.internal.MParticleApiClient;
import com.mparticle.testutils.MParticleUtils;

import java.io.IOException;
import java.net.HttpURLConnection;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSocketFactory;

public class PinningTestHelper {
    Context mContext;
    Callback mCallback;

    PinningTestHelper(Context context, String path, Callback callback) {
        mContext = context;
        mCallback = callback;
        prepareIdentityApiClient(path);
        prepareMParticleApiClient(path);
    }

    private void prepareIdentityApiClient(String path) {
        MParticleUtils.getInstance().setDefaultIdentityClient(mContext);
        com.mparticle.identity.AccessUtils.setIdentityApiClientScheme("https");
        MParticleIdentityClient apiClient = com.mparticle.identity.AccessUtils.getIdentityApiClient();
        setRequestClient(apiClient, path);
    }

    private void prepareMParticleApiClient(String path) {
        MParticleUtils.getInstance().setDefaultClient(mContext);
        com.mparticle.internal.AccessUtils.setMParticleApiClientProtocol("https");
        MParticleApiClient apiClient = com.mparticle.internal.AccessUtils.getApiClient();
        setRequestClient(apiClient, path);
    }

    private void setRequestClient(MParticleBaseClient client, final String path) {
        final BaseNetworkConnection requestHandler = client.getRequestHandler();
        client.setRequestHandler(new BaseNetworkConnection(mContext) {
            @Override
            public HttpURLConnection makeUrlRequest(MParticleBaseClientImpl.Endpoint endpoint, HttpURLConnection connection, String payload, boolean identity) throws IOException {
                try {
                    connection = requestHandler.makeUrlRequest(endpoint, connection, null, identity);
                }
                finally {
                    if (connection.getURL().toString().contains(path)) {
                        final HttpURLConnection finalConnection = connection;
                        new Handler(Looper.getMainLooper()).post(new Runnable() {
                            @Override
                            public void run() {
                                if (finalConnection instanceof HttpsURLConnection) {
                                    HttpsURLConnection sslConnection = (HttpsURLConnection) finalConnection;
                                    mCallback.onPinningApplied(isPinned(sslConnection));
                                } else {
                                    mCallback.onPinningApplied(false);
                                }
                            }
                        });
                    }
                }
                return connection;
            }
        });
    }

    /**
     * The only way I have been able to find if we are pinning certificates or not, is to test
     * whether the HttpsURLConnection is using it's origin SSLSocketFactory or not. This does not
     * tell us if it actually has certificates pinned, but rather, if it has ever been set. Not the
     * best approach, but there is not easier way, without doing some Reflection, which we should
     * eventually do
     */
    private boolean isPinned(HttpsURLConnection connection) {
        return connection.getSSLSocketFactory() != SSLSocketFactory.getDefault();
    }

    interface Callback {
        void onPinningApplied(boolean pinned);
    }
}

File: android-core/src/androidTest/kotlin/com.mparticle/networking/PinningTestHelper.kt
package com.mparticle.networking

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.mparticle.MParticle
import com.mparticle.identity.MParticleIdentityClient
import com.mparticle.internal.AccessUtils
import com.mparticle.internal.Constants
import com.mparticle.internal.MParticleApiClient
import com.mparticle.internal.MParticleApiClientImpl
import com.mparticle.internal.MParticleApiClientImpl.MPNoConfigException
import java.io.IOException
import java.net.MalformedURLException
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLSocketFactory

class PinningTestHelper internal constructor(
    var context: Context,
    path: String,
    var mCallback: (Boolean) -> Unit
) {
    private fun prepareIdentityApiClient(path: String) {
        com.mparticle.identity.AccessUtils.setDefaultIdentityApiClient(context)
        //        com.mparticle.identity.AccessUtils.setIdentityApiClientScheme("https");
        val apiClient: MParticleIdentityClient =
            com.mparticle.identity.AccessUtils.getIdentityApiClient()
        setRequestClient(apiClient, path)
    }

    private fun prepareMParticleApiClient(path: String) {
        try {
            AccessUtils.setMParticleApiClient(
                MParticleApiClientImpl(
                    MParticle.getInstance()!!.Internal().configManager,
                    context.getSharedPreferences(
                        Constants.PREFS_FILE,
                        Context.MODE_PRIVATE
                    ),
                    context
                )
            )
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: MPNoConfigException) {
            e.printStackTrace()
        }
        val apiClient: MParticleApiClient = AccessUtils.getApiClient()
        setRequestClient(apiClient, path)
    }

    private fun setRequestClient(client: MParticleBaseClient, path: String) {
        val requestHandler = client.requestHandler
        client.requestHandler = object : BaseNetworkConnection(context) {
            @Throws(IOException::class)
            override fun makeUrlRequest(
                endpoint: MParticleBaseClientImpl.Endpoint,
                connection: MPConnection,
                payload: String,
                identity: Boolean
            ): MPConnection {
                var connection = connection
                connection = try {
                    requestHandler.makeUrlRequest(endpoint, connection, null, identity)
                } finally {
                    if (connection.url.toString().contains(path)) {
                        val finalConnection = connection
                        Handler(Looper.getMainLooper()).post {
                            mCallback(
                                finalConnection.isHttps && finalConnection.sslSocketFactory != null
                            )
                        }
                    }
                }
                return connection
            }
        }
    }

    /**
     * The only way I have been able to find if we are pinning certificates or not, is to test
     * whether the HttpsURLConnection is using its origin SSLSocketFactory or not. This does not
     * tell us if it actually has certificates pinned, but rather, if it has ever been set. Not the
     * best approach, but there is no easier way, without doing some Reflection, which we should
     * eventually do.
     */
    private fun isPinned(connection: HttpsURLConnection): Boolean {
        return connection.sslSocketFactory !== SSLSocketFactory.getDefault()
    }

    interface Callback {
        fun onPinningApplied(pinned: Boolean)
    }

    init {
        prepareIdentityApiClient(path)
        prepareMParticleApiClient(path)
    }
}
